src/content_scripts/applyColorSchemeJs.js

```diff
@@ -22,109 +51,294 @@ const ADDON_FAKED_WARNING = "matchMedia has been faked/overwritten by add-on web
+    },
+    get matches() {
+        if (!checkIsMediaQueryList(this) ||
+            fakedColorStatus === COLOR_STATUS.NO_OVERWRITE
+        ) {
+            return Reflect.apply(originalMatchesGetter, this, arguments);
+        }
+        let result = evaluateMediaQuery(this.media);
+        if (result === null) {
+            return Reflect.apply(originalMatchesGetter, this, arguments);
+        }
+        mayLogFakeWarning();
+        return result;
+    },
+    get onchange() {
+        if (!checkIsMediaQueryList(this) ||
+            !hasColorMediaQuery(this.media)
+        ) {
+            return Reflect.apply(originalOnChangeGetter, this, arguments);
+        }
+        let hook = Reflect.apply(privilegedOnChangeGetter, this, arguments);
+        if (typeof hook !== 'function') {
+            return hook;
+        }
+        let func = wmHookToFunc.get(hook);
+        if (typeof func !== 'function') {
+            console.error('[website-dark-mode-switcher] someone called "get onchange" on an unknown MediaQueryList!');
+            return null;
+        }
+        return func;
+    },
+    set onchange(func) {
+        if (!checkIsMediaQueryList(this) ||
+            typeof func !== 'function' ||
+            !hasColorMediaQuery(this.media)
+        ) {
+            // eslint-disable-next-line no-setter-return
+            return Reflect.apply(originalOnChangeSetter, this, arguments);
+        }
+        let oldHook = Reflect.apply(privilegedOnChangeGetter, this, arguments);
+        if (typeof oldHook === 'function') {
+            let oldFunc = wmHookToFunc.get(oldHook);
+            if (typeof oldFunc !== 'function') {
+                console.error('[website-dark-mode-switcher] someone called "set onchange" on an unknown MediaQueryList!');
+                return;
+            }
+            trackOffListener(oldFunc, this, true);
+        }
+        let hook = trackOnListener(func, this, true);
+        // eslint-disable-next-line no-setter-return
+        return Reflect.apply(originalOnChangeSetter, this, [hook]);
+    },
+
+    addEventListener(type, listener, options) {
+        if (!checkIsMediaQueryList(this) ||
+            type !== 'change' ||
+            typeof listener !== 'function' ||
+            !hasColorMediaQuery(this.media)
+        ) {
+            return Reflect.apply(originalAddEventListener, this, arguments);
+        }
+        let hook = trackOnListener(listener, this, false);
+        return Reflect.apply(originalAddEventListener, this, ['change', hook, options]);
+    },
+    removeEventListener(type, listener, options) {
+        if (!checkIsMediaQueryList(this) ||
+            type !== 'change' ||
+            typeof listener !== 'function' ||
+            !hasColorMediaQuery(this.media)
+        ) {
+            return Reflect.apply(originalRemoveEventListener, this, arguments);
+        }
+        let hook = trackOffListener(listener, this, false);
+        if (!hook) {
```

Comment:
Wanna do a strict function type-check here, too?

Reply:
Hooks are something we created, as opposed listeners that are passed from untrusted code, I'm not as paranoid there.

Reply:
Okay, fine. Just saw it in the code, so yeah. :slightly_smiling_face: